dan_layer/engine_types/src/substate.rs

```diff
@@ -351,33 +351,45 @@ impl FromStr for SubstateId {
                 Ok(SubstateId::Component(addr))
             },
             Some(("resource", addr)) => {
-                match addr.split_once(' ') {
-                    Some((resource_str, addr)) => match addr.split_once('_') {
-                        // resource_xxxx nft_xxxxx
-                        Some(("nft", _)) => {
-                            let nft_address =
-                                NonFungibleAddress::from_str(s).map_err(|e| InvalidSubstateIdFormat(e.to_string()))?;
-                            Ok(SubstateId::NonFungible(nft_address))
-                        },
-                        // resource_xxxx index_
-                        Some(("index", index_str)) => {
-                            let resource_addr = ResourceAddress::from_hex(resource_str)
-                                .map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                            let index = u64::from_str(index_str).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                            Ok(SubstateId::NonFungibleIndex(NonFungibleIndexAddress::new(
-                                resource_addr,
-                                index,
-                            )))
-                        },
-                        _ => Err(InvalidSubstateIdFormat(s.to_string())),
-                    },
-                    // resource_xxxx
-                    None => {
-                        let addr =
-                            ResourceAddress::from_hex(addr).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                        Ok(SubstateId::Resource(addr))
-                    },
+                if addr.split_once('_').is_none() {
+                    // resource_xxxxx
+                    let addr = ResourceAddress::from_hex(addr).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
+                    return Ok(SubstateId::Resource(addr));
                 }
+                Err(InvalidSubstateIdFormat("Missing resource address".to_string()))
+            },
+            Some(("nft", rest)) => {
+                // nft_{resource_addr}_{id_type}_{id}
+                let (resource, nft_rest) = rest
+                    .split_once('_')
+                    .ok_or_else(|| InvalidSubstateIdFormat(s.to_string()))?;
+
+                let resource_addr =
+                    ResourceAddress::from_hex(resource).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
+
+                let (id_type, id) = nft_rest
+                    .split_once('_')
+                    .ok_or_else(|| InvalidSubstateIdFormat(s.to_string()))?;
+
+                let nft_id = NonFungibleId::try_from_canonical_string(&format!("{id_type}_{id}"))
+                    .map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
```

Comment:
try_from_canonical_string will already error. There's no need to split and rejoin the string (which allocates a string on the heap)
```suggestion
            let nft_id = NonFungibleId::try_from_canonical_string(nft_rest))
                    .map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
```